flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/sessionjob/SessionJobReconcilerTest.java

```diff
@@ -409,24 +417,143 @@ public void testTriggerSavepoint() throws Exception {
+                "savepoint_trigger_1",
                 sp1SessionJob.getStatus().getJobStatus().getSavepointInfo().getTriggerId());
 
         sp1SessionJob.getStatus().getJobStatus().getSavepointInfo().resetTrigger();
-        ReconciliationUtils.updateLastReconciledSavepointTriggerNonce(
-                sp1SessionJob.getStatus().getJobStatus().getSavepointInfo(), sp1SessionJob);
+        ReconciliationUtils.updateLastReconciledSnapshotTriggerNonce(
+                sp1SessionJob.getStatus().getJobStatus().getSavepointInfo(),
+                sp1SessionJob,
+                SAVEPOINT);
 
         // don't trigger when nonce is cleared
         sp1SessionJob.getSpec().getJob().setSavepointTriggerNonce(null);
         reconciler.reconcile(sp1SessionJob, readyContext);
-        assertFalse(SavepointUtils.savepointInProgress(sp1SessionJob.getStatus().getJobStatus()));
+        assertFalse(SnapshotUtils.savepointInProgress(sp1SessionJob.getStatus().getJobStatus()));
+    }
+
+    @Test
+    public void testTriggerCheckpoint() throws Exception {
+        FlinkSessionJob sessionJob = TestUtils.buildSessionJob();
+        assertFalse(SnapshotUtils.checkpointInProgress(getJobStatus(sessionJob)));
+
+        var readyContext = TestUtils.createContextWithReadyFlinkDeployment();
+        reconciler.reconcile(sessionJob, readyContext);
+        verifyAndSetRunningJobsToStatus(
+                sessionJob, JobState.RUNNING, RECONCILING.name(), null, flinkService.listJobs());
+
+        assertFalse(SnapshotUtils.checkpointInProgress(getJobStatus(sessionJob)));
+
+        // trigger checkpoint
+        var sp1SessionJob = ReconciliationUtils.clone(sessionJob);
+
+        // do not trigger checkpoint if nonce is null
+        reconciler.reconcile(sp1SessionJob, readyContext);
+        assertFalse(SnapshotUtils.checkpointInProgress(getJobStatus(sp1SessionJob)));
+
+        getJobSpec(sp1SessionJob).setCheckpointTriggerNonce(2L);
+        getJobStatus(sp1SessionJob).setState(CREATED.name());
+        reconciler.reconcile(sp1SessionJob, readyContext);
+        // do not trigger checkpoint if job is not running
+        assertFalse(SnapshotUtils.checkpointInProgress(getJobStatus(sp1SessionJob)));
+
+        getJobStatus(sp1SessionJob).setState(RUNNING.name());
+
+        reconciler.reconcile(sp1SessionJob, readyContext);
+        assertTrue(SnapshotUtils.checkpointInProgress(getJobStatus(sp1SessionJob)));
+
+        // the last reconcile nonce updated
+        assertNull(getReconciledJobSpec(sp1SessionJob).getCheckpointTriggerNonce());
+
+        // don't trigger new checkpoint when checkpoint is in progress
+        getJobSpec(sp1SessionJob).setCheckpointTriggerNonce(3L);
+        reconciler.reconcile(sp1SessionJob, readyContext);
+        assertEquals("checkpoint_trigger_0", getCheckpointInfo(sp1SessionJob).getTriggerId());
+        /*
+            TODO: this section needs to be reintroduced in case the LAST_STATE optimization gets
+             added
```

Comment:
Do we want to keep this? Not sure it will be useful to somebody else when we add the feature later in time.

Reply:
I mostly left it for the review discussion. @gyfora what do you think, is is a worthy optimization in your opinion?

Reply:
Sorry, what optimisation are we talking about here?

Reply:
To block upgrades until checkpointing is complete if the LATEST_STATE is enabled 
https://a1350286.slack.com/archives/C02V70Y3VAQ/p1689850156914949?thread_ts=1689781790.906109&cid=C02V70Y3VAQ

Reply:
I would add a JIRA and remove this comment.

Reply:
👍  https://issues.apache.org/jira/browse/FLINK-32719
Removed.